brushfire-core/src/main/scala/com/stripe/brushfire/TreeTraversal.scala

```diff
@@ -0,0 +1,149 @@
+   * `true` predicates. However, when multiple predicates are `true` in a
+   * [[SplitNode]], the actual choice of which one gets chosen is left to the
+   * particular implementation of `TreeTraversal`.
+   *
+   * @param tree the decision tree to search in
+   * @param row  the row/instance we're trying to match with a leaf node
+   * @return the leaf nodes that best match the row
+   */
+  def find(tree: AnnotatedTree[K, V, T, A], row: Map[K, V]): Stream[LeafNode[K, V, T, A]] =
+    find(tree.root, row)
+
+  /**
+   * Find the [[LeafNode]] that best fits `row` and is a descendant of `init`.
+   * Generally, the path from `init` to the resulting leaf node will be along
+   * *only* `true` predicates. However, when multiple predicates are `true` in
+   * a [[SplitNode]], the actual choice of which one gets chosen is left to the
+   * particular implementation of `TreeTraversal`.
+   *
+   * @param init the initial node to start from
+   * @param row  the row/instance we're trying to match with a leaf node
+   * @return the leafs node that match the row
+   */
+  def find(node: Node[K, V, T, A], row: Map[K, V]): Stream[LeafNode[K, V, T, A]]
+}
+
+object TreeTraversal {
+
+  def find[K, V, T, A](tree: AnnotatedTree[K, V, T, A], row: Map[K, V])(implicit traversal: TreeTraversal[K, V, T, A]): Stream[LeafNode[K, V, T, A]] =
+    traversal.find(tree, row)
+
+  /**
+   * Performs a depth-first traversal of the tree, returning all matching leaf
+   * nodes.
+   */
+  implicit def depthFirst[K, V, T, A]: TreeTraversal[K, V, T, A] =
+    DepthFirstTreeTraversal(identity)
+
+  /**
+   * A depth first search for matching leaves, randomly choosing the order of
+   * child candidate nodes to traverse at each step. Since it is depth-first,
+   * after a node is chosen to be traversed, all of the matching leafs that
+   * descend from that node are traversed before moving onto the node's
+   * sibling.
+   */
+  def randomDepthFirst[K, V, T, A](rng: Random = Random): TreeTraversal[K, V, T, A] =
+    DepthFirstTreeTraversal(rng.shuffle(_))
+
+  /**
+   * A depth-first search for matching leaves, where the candidate child nodes
+   * for a given parent node are traversed in reverse order of their
+   * annotations. This means that if we have multiple valid candidate children,
+   * we will traverse the child with the largest annotation first.
+   */
+  def weightedDepthFirst[K, V, T, A: Ordering]: TreeTraversal[K, V, T, A] =
+    DepthFirstTreeTraversal(_.sortBy(_.annotation)(Ordering[A].reverse))
+
+  /**
+   * A depth-first search for matching leaves, where the candidate child leaves
+   * of a parent node are randomly shuffled, but with nodes with higher weight
+   * being given a higher probability of being ordered earlier. This is
+   * basically a mix between [[randomDepthFirst]] and [[weightedDepthFirst]].
+   */
+  def probabilisticWeightedDepthFirst[K, V, T, A <% Double](rng: Random = Random): TreeTraversal[K, V, T, A] =
```

Comment:
I kind of want to make all the `find` methods take a `Random` (with default) instead of having the traversal itself trap a random. It's a leaky abstraction, but would make things saner down the road.